### PR TITLE
Fix NSData tests on reference platform

### LINKED ISSUE #748: NSData long description tests fail on OSX
Seems to be slightly off from OSX long descriptions - interesting that this happened in ReferenceFoundation:

[ RUN      ] NSData.LongDebugDescription
Foundation/ReferenceFoundation/TestNSData.mm:160: Failure
Value of: expected
  Actual: \<ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ... ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff\>
Expected: data.debugDescription
Which is: <OS_dispatch_data: data[0x7fae98c1f760] = { leaf, size = 100000, buf = 0x10f144000 }>
[  FAILED  ] NSData.LongDebugDescription (3 ms)
[ RUN      ] NSData.EdgeDebugDescription
Foundation/ReferenceFoundation/TestNSData.mm:187: Failure
Value of: expected
  Actual: \<ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ... ffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ff\>
Expected: data.debugDescription
Which is: \<ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ff\>
[  FAILED  ] NSData.EdgeDebugDescription (2 ms)
[ RUN      ] NSData.EdgeNoCopyDescription
Foundation/ReferenceFoundation/TestNSData.mm:214: Failure
Value of: expected
  Actual: \<ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ... ffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ff\>
Expected: data.debugDescription
Which is: \<ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ff\>
[  FAILED  ] NSData.EdgeNoCopyDescription (2 ms)

### PR DESCRIPTION
**Pull request by aballway**
Fix NSData description so it returns the same values as reference platform's version, and change initWithBase64Encoding to return an empty NSData object when given invalid input rather than nil as documentaiton suggests

Fixes #748 and #749 NSData.Base64DecodeWithOptions test failure

### CODE REVIEW

**Comment by aballway** [comment]
NSData.LongDebugDescription seems to fail because the reference platform uses a different class (NSObject) in the NSData class cluster when the size of the data is >= 2^16 - 9 bytes (a little under 64kb) on my test machine, possibly different sizes depending on the test platform. To fix this test we can either A: create the OS_dispatch_data protocol for NSObject, which will probably require quite a lot of work or B: lower the size of the test data so we do not encounter this issue. Because this seems to be more an implementation difference that should only come up during logging than an actual functionality one, I went with option B.

File: Frameworks/Foundation/NSData.mm [hunk 1]
```diff
@@ -447,7 +447,12 @@ - (instancetype)initWithBase64EncodedData:(NSData*)base64Data options:(NSDataBas
  @Status Interoperable
 */
 - (instancetype)initWithBase64Encoding:(NSString*)base64String {
-    return [self initWithBase64EncodedString:base64String options:0];
+    // Documentation states that invalid input should return nil, but reference platform tests show it returns an empty NSData object
+    auto ret = [self initWithBase64EncodedString:base64String options:0];
+    if (!ret) {
+        ret = [self initWithBase64EncodedString:@"" options:0];
+    }
+    return ret;
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSData.mm]: !!! This is alarmingly different behaviour, but I can verify.

**Comment by DHowett-MSFT** [comment]
:shipit:

File: tests/unittests/Foundation/NSDataTests.mm [hunk 2]
```diff
@@ -97,9 +97,9 @@ void testDecode(NSString* base64String, NSString* expectedDecode, BOOL ignoreUnk
 }
 
 TEST(NSData, Base64DecodeWithOptions) {
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSDataTests.mm]: >odeWithOp [](start = 22, length = 9)

This test says `WithOptions`, but it is using the option-free variety.
Should we add another test for the non-option-free variety, which DOES return nil for invalid data?

File: Frameworks/Foundation/NSData.mm [hunk 3]
```diff
@@ -447,7 +447,12 @@ - (instancetype)initWithBase64EncodedData:(NSData*)base64Data options:(NSDataBas
  @Status Interoperable
 */
 - (instancetype)initWithBase64Encoding:(NSString*)base64String {
-    return [self initWithBase64EncodedString:base64String options:0];
+    // Documentation states that invalid input should return nil, but reference platform tests show it returns an empty NSData object
+    auto ret = [self initWithBase64EncodedString:base64String options:0];
+    if (!ret) {
+        ret = [self initWithBase64EncodedString:@"" options:0];
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSData.mm]: >[self initWithBase64EncodedString:@"" options:0]; [](start = 14, length = 49)

[self init] should be sufficient here.

**Comment by rajsesh-msft** [comment]
:shipit:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm [hunk 4]
```diff
@@ -106,7 +106,7 @@
     ASSERT_OBJCEQ(data.debugDescription, expected);
 }
 
-TEST(NSData, LimitDebugDescription) {
+TEST(NSData, LongDebugDescription) {
```
Comment (ms-jihua) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: Keep this as 'LimitDebugDescription'.

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm [hunk 5]
```diff
@@ -133,34 +133,7 @@
     ASSERT_OBJCEQ(data.debugDescription, expected);
 }
 
-TEST(NSData, LongDebugDescription) {
```
Comment (ms-jihua) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: Make this an OSX_DISABLED_TEST instead, with an explanation as to why.

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm [hunk 6]
```diff
@@ -133,34 +133,7 @@
...
-        @"ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff "
-        @"ffffffff ffffffff ffffffff ffffffff ffffffff>";
-
-    std::vector<uint8_t> bytes(100000, 0xff);
-    NSData* data = [NSData dataWithBytes:bytes.data() length:bytes.size()];
-    ASSERT_OBJCEQ(data.debugDescription, expected);
-}
-
-TEST(NSData, EdgeDebugDescription) {
+TEST(NSData, NoCopyDescription) {
```
Comment (ms-jihua) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: Keep both EdgeDebugDescription and EdgeNoCopyDescription - there's a difference in the initializers they use.

To get around the behavior you mentioned, change the number of bytes to 513 instead of 1025 - the behavior they're concerned with validating is the 'ff>' at the end, the remainder byte.

Keep the 1025 code in a comment (probably not the expected value though, seems unnecessary), with an explanation as to why we changed it from 
[https://github.com/apple/swift-corelibs-foundation/blob/master/TestFoundation/TestNSData.swift](https://github.com/apple/swift-corelibs-foundation/blob/master/TestFoundation/TestNSData.swift)

File: Frameworks/Foundation/NSData.mm [hunk 7]
```diff
@@ -609,27 +614,14 @@ - (NSString*)description {
...
-    const int tmpBufSize = 16 + lengthToUse * 2 + (lengthToUse / 4) + (length > 1024 ? 4 : 0);
+    const int tmpBufSize = 16 + length * 2 + (length / 4);
 
     std::vector<char> tmpBuf(tmpBufSize);
     int tmpBufLen = 0;
 
     tmpBuf[tmpBufLen++] = '<';
 
     for (auto i = 0; i < length;) {
-        // skip the middle portion, if length > 1024, and go directly to the last 512 bytes
```
Comment (ms-jihua) [inline Frameworks/Foundation/NSData.mm]: You might as well keep this behavior - iirc it matches iOS, and it doesn't hurt us to trim overly-long descriptions anyway.

**Comment by ms-jihua** [comment]
:clock1:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm (hunk 6 above)
Comment (aballway) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: WRT these tests, wouldn't it make more sense to match the OSX method functionality and remove what would be redundant tests?

---
In reply to: [73397446](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73397446) [](ancestors = 73397446)

File: tests/unittests/Foundation/NSDataTests.mm [hunk 8]
```diff
@@ -96,11 +96,17 @@ void testDecode(NSString* base64String, NSString* expectedDecode, BOOL ignoreUnk
                       base64EncodedStringWithOptions:NSDataBase64Encoding76CharacterLineLength | NSDataBase64EncodingEndLineWithLineFeed]);
 }
 
-TEST(NSData, Base64DecodeWithOptions) {
-    // Invalid input should return nil
+TEST(NSData, Base64EncodingWithoutOptions) {
```
Comment (ms-jihua) [inline tests/unittests/Foundation/NSDataTests.mm]: >Base64EncodingWithoutOption [](start = 13, length = 27)

This name doesn't really describe this test.

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm (hunk 6 above)
Comment (ms-jihua) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: They're not redundant - look closely: the initializers are different, and the latter has an extra assert.

---
In reply to: [73406834](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73406834) [](ancestors = 73406834,73397446)

**Comment by msft-Jeyaram** [comment]
:shipit:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm (hunk 6 above)
Comment (aballway) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: With this we're keeping a test with dataWithBytes and one with dataWithBytesNoCopy and the extra assert, but the other two dataWithBytes tests only tested the description shortening functionality, which isn't how it works on OSX.  

---
In reply to: [73432695](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73432695) [](ancestors = 73432695,73406834,73397446)

File: Frameworks/Foundation/NSData.mm [hunk 9]
```diff
@@ -447,7 +447,12 @@ - (instancetype)initWithBase64EncodedData:(NSData*)base64Data options:(NSDataBas
  @Status Interoperable
 */
 - (instancetype)initWithBase64Encoding:(NSString*)base64String {
-    return [self initWithBase64EncodedString:base64String options:0];
+    // Documentation states that invalid input should return nil, but reference platform tests show it returns an empty NSData object
+    auto ret = [self initWithBase64EncodedString:base64String options:0];
```
Comment (bbowman) [inline Frameworks/Foundation/NSData.mm]: >initWithBase64EncodedString [](start = 21, length = 27)

But the init method DOES return nil on the reference platform?!

File: Frameworks/Foundation/NSData.mm (hunk 7 above)
Comment (bbowman) [inline Frameworks/Foundation/NSData.mm]: IIRC this was needed / confirmed to be used. Why did you remove it?

---
In reply to: [73397672](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73397672) [](ancestors = 73397672)

**Comment by bbowman** [comment]
:shipit:

File: Frameworks/Foundation/NSData.mm (hunk 7 above)
Comment (aballway) [inline Frameworks/Foundation/NSData.mm]: When testing on OSX the tests were failing because it did not do this, it just printed out everything

---
In reply to: [73448519](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73448519) [](ancestors = 73448519,73397672)

Comment (ms-jihua) [inline Frameworks/Foundation/NSData.mm]: aballway confirmed that this doesn't repro on iOS - apparently this is a swift-only feature?

---
In reply to: [73544573](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73544573) [](ancestors = 73544573,73448519,73397672)

File: tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm (hunk 6 above)
Comment (ms-jihua) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSData.mm]: Discussed offline, we're gonna keep one that ends in ffffffff> (LimitDebugDescription?) for completeness, and that should be good.

---
In reply to: [73435499](https://github.com/Microsoft/WinObjC/pull/798#discussion_r73435499) [](ancestors = 73435499,73432695,73406834,73397446)
